pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+	///
+	/// `frame_system::Config` should always be included.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + timestamp::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Address which holds the customer funds.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Balances Pallet
+		type NativeCurrency: Currency<Self::AccountId>
+			+ ReservableCurrency<Self::AccountId>
+			+ LockableCurrency<Self::AccountId>;
+
+		type Public: Clone
+			+ PartialEq
+			+ IdentifyAccount<AccountId = Self::AccountId>
+			+ Debug
+			+ parity_scale_codec::Codec
+			+ Ord
+			+ scale_info::TypeInfo;
+
+		/// A matching `Signature` type.
+		type Signature: Verify<Signer = Self::Public>
+			+ Clone
+			+ PartialEq
+			+ Debug
+			+ parity_scale_codec::Codec
+			+ scale_info::TypeInfo;
+
+		/// Governance Origin
+		type GovernanceOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+	}
+
+	// Simple declaration of the `Pallet` type. It is placeholder we use to implement traits and
+	// method.
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// The extrinsic will be used to start a new reward cycle
+		/// # Parameters
+		/// * `origin`: The donor who wants to start the reward cycle
+		/// * `start_block`: The block from which reward distribution will start
+		/// * `end_block`: The block at which last rewards will be distributed
+		/// * `initial_percentage`: The percentage of rewards that can be claimed at start block
+		/// * `reward_id`: The reward id
+		#[pallet::weight(10_000)]
+		pub fn create_reward_cycle(
+			origin: OriginFor<T>,
+			start_block: T::BlockNumber,
+			end_block: T::BlockNumber,
+			initial_percentage: u32,
+			reward_id: u32,
+		) -> DispatchResult {
+			//check to ensure governance
+			T::GovernanceOrigin::ensure_origin(origin.clone())?;
+
+			//check to ensure no duplicate id gets added
+			ensure!(!<InitializeRewards<T>>::contains_key(reward_id), Error::<T>::DuplicateId);
+
+			//check to ensure start block greater than end block
+			ensure!(start_block < end_block, Error::<T>::InvalidBlocksRange);
+
+			//ensure percentage range is valid
+			ensure!(
+				initial_percentage <= 100 && initial_percentage > 0,
+				Error::<T>::InvalidInitialPercentage
+			);
+
+			let reward_info = RewardInfo { start_block, end_block, initial_percentage };
+
+			//inserting reward info into the storage
+			<InitializeRewards<T>>::insert(reward_id, reward_info);
+
+			Self::deposit_event(Event::RewardCycleCreated { start_block, end_block, reward_id });
+
+			Ok(())
+		}
+
+		///The extrinsic will transfer and lock users rewards into users account
+		/// # Parameters,
+		/// * `origin`: The users address which has been mapped to reward id
+		/// * `reward_id`: Reward id
+		#[pallet::weight(10_000)]
```

Comment:
No weights through benchmarks

Reply:
Issue https://github.com/Polkadex-Substrate/Polkadex/issues/629 opened for it.